// @camb-ai/sdk/dist/api/resources/streaming/client/Client.js
// This file was auto-generated by Fern from our API Definition.
import { normalizeClientOptionsWithAuth } from "../../../../BaseClient.js";
import { mergeHeaders } from "../../../../core/headers.js";
import * as core from "../../../../core/index.js";
import * as environments from "../../../../environments.js";
import { handleNonStatusCodeError } from "../../../../errors/handleNonStatusCodeError.js";
import * as errors from "../../../../errors/index.js";
import * as CambApi from "../../../index.js";
export class StreamingClient {
    constructor(options) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    /**
     * @param {CambApi.CreateStreamRequestPayload} request
     * @param {StreamingClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.streaming.createStream({
     *         voices: [1],
     *         dictionaries: [1],
     *         config: {},
     *         source_stream: {
     *             language: 1,
     *             url: "url"
     *         },
     *         target_streams: [{
     *                 languages: [1],
     *                 url: "url",
     *                 type: 1
     *             }]
     *     })
     */
    createStream(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__createStream(request, requestOptions));
    }
    async __createStream(request, requestOptions) {
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "stream"),
            method: "POST",
            headers: _headers,
            contentType: "application/json",
            queryParameters: requestOptions?.queryParams,
            requestType: "json",
            body: request,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/stream");
    }
    /**
     * @param {CambApi.GetStreamResultStreamStreamIdGetRequest} request
     * @param {StreamingClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.streaming.getStreamResult({
     *         stream_id: 1
     *     })
     */
    getStreamResult(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getStreamResult(request, requestOptions));
    }
    async __getStreamResult(request, requestOptions) {
        const { stream_id: streamId, run_id: runId } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `stream/${core.url.encodePathParam(streamId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/stream/{stream_id}");
    }
    /**
     * @param {CambApi.DestroyStreamStreamStreamIdDeleteRequest} request
     * @param {StreamingClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.streaming.destroyStream({
     *         stream_id: 1
     *     })
     */
    destroyStream(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__destroyStream(request, requestOptions));
    }
    async __destroyStream(request, requestOptions) {
        const { stream_id: streamId, run_id: runId } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `stream/${core.url.encodePathParam(streamId)}`),
            method: "DELETE",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "DELETE", "/stream/{stream_id}");
    }
    /**
     * @param {CambApi.UpdateStreamDataRequestPayload} request
     * @param {StreamingClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.streaming.patchStreamData({
     *         stream_id: 1
     *     })
     */
    patchStreamData(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__patchStreamData(request, requestOptions));
    }
    async __patchStreamData(request, requestOptions) {
        const { stream_id: streamId, run_id: runId, ..._body } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `stream/${core.url.encodePathParam(streamId)}`),
            method: "PATCH",
            headers: _headers,
            contentType: "application/json",
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            requestType: "json",
            body: _body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "PATCH", "/stream/{stream_id}");
    }
    /**
     * @param {CambApi.GetProbeStreamStreamProbePostRequest} request
     * @param {StreamingClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.streaming.getProbeStream({
     *         body: {
     *             url: "url"
     *         }
     *     })
     */
    getProbeStream(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getProbeStream(request, requestOptions));
    }
    async __getProbeStream(request, requestOptions) {
        const { run_id: runId, body: _body } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "stream/probe"),
            method: "POST",
            headers: _headers,
            contentType: "application/json",
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            requestType: "json",
            body: _body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/stream/probe");
    }
}
