// @camb-ai/sdk/dist/api/resources/textToVoice/client/Client.js
// This file was auto-generated by Fern from our API Definition.
import { normalizeClientOptionsWithAuth } from "../../../../BaseClient.js";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../../core/headers.js";
import * as core from "../../../../core/index.js";
import * as environments from "../../../../environments.js";
import { handleNonStatusCodeError } from "../../../../errors/handleNonStatusCodeError.js";
import * as errors from "../../../../errors/index.js";
import * as CambApi from "../../../index.js";
export class TextToVoiceClient {
    constructor(options) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    /**
     * @param {CambApi.CreateTextToVoiceRequestPayload} request
     * @param {TextToVoiceClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.textToVoice.createTextToVoice({
     *         text: "text",
     *         voice_description: "voice_description"
     *     })
     */
    createTextToVoice(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__createTextToVoice(request, requestOptions));
    }
    async __createTextToVoice(request, requestOptions) {
        const { traceparent, ..._body } = request;
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, mergeOnlyDefinedHeaders({ traceparent: traceparent !== undefined ? traceparent : undefined }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "text-to-voice"),
            method: "POST",
            headers: _headers,
            contentType: "application/json",
            queryParameters: requestOptions?.queryParams,
            requestType: "json",
            body: _body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/text-to-voice");
    }
    /**
     * @param {CambApi.GetTextToVoiceStatusTextToVoiceTaskIdGetRequest} request
     * @param {TextToVoiceClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.textToVoice.getTextToVoiceStatus({
     *         task_id: "task_id"
     *     })
     */
    getTextToVoiceStatus(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTextToVoiceStatus(request, requestOptions));
    }
    async __getTextToVoiceStatus(request, requestOptions) {
        const { task_id: taskId, run_id: runId, traceparent } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, mergeOnlyDefinedHeaders({ traceparent: traceparent !== undefined ? traceparent : undefined }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `text-to-voice/${core.url.encodePathParam(taskId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/text-to-voice/{task_id}");
    }
    /**
     * @param {CambApi.GetTextToVoiceResultTextToVoiceResultRunIdGetRequest} request
     * @param {TextToVoiceClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.textToVoice.getTextToVoiceResult({
     *         run_id: 1
     *     })
     */
    getTextToVoiceResult(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTextToVoiceResult(request, requestOptions));
    }
    async __getTextToVoiceResult(request, requestOptions) {
        const { run_id: runId } = request;
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `text-to-voice-result/${core.url.encodePathParam(runId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: requestOptions?.queryParams,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/text-to-voice-result/{run_id}");
    }
}
