// @camb-ai/sdk/dist/api/resources/languages/client/Client.js
// This file was auto-generated by Fern from our API Definition.
import { normalizeClientOptionsWithAuth } from "../../../../BaseClient.js";
import { mergeHeaders } from "../../../../core/headers.js";
import * as core from "../../../../core/index.js";
import * as environments from "../../../../environments.js";
import { handleNonStatusCodeError } from "../../../../errors/handleNonStatusCodeError.js";
import * as errors from "../../../../errors/index.js";
import * as CambApi from "../../../index.js";
export class LanguagesClient {
    constructor(options) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    /**
     * @param {CambApi.GetSourceLanguagesSourceLanguagesGetRequest} request
     * @param {LanguagesClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.languages.getSourceLanguages()
     */
    getSourceLanguages(request = {}, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getSourceLanguages(request, requestOptions));
    }
    async __getSourceLanguages(request = {}, requestOptions) {
        const { run_id: runId } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "source-languages"),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/source-languages");
    }
    /**
     * @param {CambApi.GetTargetLanguagesTargetLanguagesGetRequest} request
     * @param {LanguagesClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.languages.getTargetLanguages()
     */
    getTargetLanguages(request = {}, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTargetLanguages(request, requestOptions));
    }
    async __getTargetLanguages(request = {}, requestOptions) {
        const { run_id: runId } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "target-languages"),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/target-languages");
    }
}
