// @camb-ai/sdk/dist/BaseClient.js
// This file was auto-generated by Fern from our API Definition.
import { HeaderAuthProvider } from "./auth/HeaderAuthProvider.js";
import { mergeHeaders } from "./core/headers.js";
import * as core from "./core/index.js";
export function normalizeClientOptions(options) {
    // Validate provider configuration
    const hasApiKey = options.apiKey != null && options.apiKey !== "";
    const hasProviderConfig = options.ttsProvider != null && options.providerParams != null;
    if (!hasApiKey && !hasProviderConfig) {
        throw new Error("Please provide either 'apiKey' or both 'ttsProvider' and 'providerParams'.");
    }
    const headers = mergeHeaders({
        "X-Fern-Language": "JavaScript",
        "X-Fern-Runtime": core.RUNTIME.type,
        "X-Fern-Runtime-Version": core.RUNTIME.version,
    }, options?.headers);
    return {
        ...options,
        logging: core.logging.createLogger(options?.logging),
        headers,
        ttsProvider: options.ttsProvider,
        providerParams: options.providerParams,
    };
}
export function normalizeClientOptionsWithAuth(options) {
    const normalized = normalizeClientOptions(options);
    const normalizedWithNoOpAuthProvider = withNoOpAuthProvider(normalized);
    normalized.authProvider ?? (normalized.authProvider = new HeaderAuthProvider(normalizedWithNoOpAuthProvider));
    return normalized;
}
function withNoOpAuthProvider(options) {
    return {
        ...options,
        authProvider: new core.NoOpAuthProvider(),
    };
}
