// @camb-ai/sdk/dist/Client.js
// This file was auto-generated by Fern from our API Definition.
import { AudioSeparationClient } from "./api/resources/audioSeparation/client/Client.js";
import { DeprecatedStreamingClient } from "./api/resources/deprecatedStreaming/client/Client.js";
import { DictionariesClient } from "./api/resources/dictionaries/client/Client.js";
import { DubClient } from "./api/resources/dub/client/Client.js";
import { FoldersClient } from "./api/resources/folders/client/Client.js";
import { LanguagesClient } from "./api/resources/languages/client/Client.js";
import { ProjectSetupClient } from "./api/resources/projectSetup/client/Client.js";
import { StoryClient } from "./api/resources/story/client/Client.js";
import { StreamingClient } from "./api/resources/streaming/client/Client.js";
import { TextToAudioClient } from "./api/resources/textToAudio/client/Client.js";
import { TextToSpeechClient } from "./api/resources/textToSpeech/client/Client.js";
import { TextToVoiceClient } from "./api/resources/textToVoice/client/Client.js";
import { TranscriptionClient } from "./api/resources/transcription/client/Client.js";
import { TranslatedStoryClient } from "./api/resources/translatedStory/client/Client.js";
import { TranslatedTtsClient } from "./api/resources/translatedTts/client/Client.js";
import { TranslationClient } from "./api/resources/translation/client/Client.js";
import { VoiceCloningClient } from "./api/resources/voiceCloning/client/Client.js";
import { normalizeClientOptionsWithAuth } from "./BaseClient.js";
import { mergeHeaders } from "./core/headers.js";
import * as core from "./core/index.js";
import * as environments from "./environments.js";
import { handleNonStatusCodeError } from "./errors/handleNonStatusCodeError.js";
import * as errors from "./errors/index.js";
export class CambClient {
    constructor(options) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    get audioSeparation() {
        return (this._audioSeparation ?? (this._audioSeparation = new AudioSeparationClient(this._options)));
    }
    get dub() {
        return (this._dub ?? (this._dub = new DubClient(this._options)));
    }
    get folders() {
        return (this._folders ?? (this._folders = new FoldersClient(this._options)));
    }
    get languages() {
        return (this._languages ?? (this._languages = new LanguagesClient(this._options)));
    }
    get story() {
        return (this._story ?? (this._story = new StoryClient(this._options)));
    }
    get translatedStory() {
        return (this._translatedStory ?? (this._translatedStory = new TranslatedStoryClient(this._options)));
    }
    get textToAudio() {
        return (this._textToAudio ?? (this._textToAudio = new TextToAudioClient(this._options)));
    }
    get textToVoice() {
        return (this._textToVoice ?? (this._textToVoice = new TextToVoiceClient(this._options)));
    }
    get textToSpeech() {
        return (this._textToSpeech ?? (this._textToSpeech = new TextToSpeechClient(this._options)));
    }
    get translation() {
        return (this._translation ?? (this._translation = new TranslationClient(this._options)));
    }
    get transcription() {
        return (this._transcription ?? (this._transcription = new TranscriptionClient(this._options)));
    }
    get translatedTts() {
        return (this._translatedTts ?? (this._translatedTts = new TranslatedTtsClient(this._options)));
    }
    get streaming() {
        return (this._streaming ?? (this._streaming = new StreamingClient(this._options)));
    }
    get voiceCloning() {
        return (this._voiceCloning ?? (this._voiceCloning = new VoiceCloningClient(this._options)));
    }
    get dictionaries() {
        return (this._dictionaries ?? (this._dictionaries = new DictionariesClient(this._options)));
    }
    get projectSetup() {
        return (this._projectSetup ?? (this._projectSetup = new ProjectSetupClient(this._options)));
    }
    get deprecatedStreaming() {
        return (this._deprecatedStreaming ?? (this._deprecatedStreaming = new DeprecatedStreamingClient(this._options)));
    }
    getSwaggerDocsDocsGet(requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getSwaggerDocsDocsGet(requestOptions));
    }
    async __getSwaggerDocsDocsGet(requestOptions) {
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "docs"),
            method: "GET",
            headers: _headers,
            queryParameters: requestOptions?.queryParams,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            throw new errors.CambApiError({
                statusCode: _response.error.statusCode,
                body: _response.error.body,
                rawResponse: _response.rawResponse,
            });
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/docs");
    }
    /**
     * @param {CambClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @example
     *     await client.getRedocDocsRedocsGet()
     */
    getRedocDocsRedocsGet(requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getRedocDocsRedocsGet(requestOptions));
    }
    async __getRedocDocsRedocsGet(requestOptions) {
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "redocs"),
            method: "GET",
            headers: _headers,
            queryParameters: requestOptions?.queryParams,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            throw new errors.CambApiError({
                statusCode: _response.error.statusCode,
                body: _response.error.body,
                rawResponse: _response.rawResponse,
            });
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/redocs");
    }
    /**
     * @param {CambClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @example
     *     await client.getOpenapiSchemaOpenapiJsonGet()
     */
    getOpenapiSchemaOpenapiJsonGet(requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getOpenapiSchemaOpenapiJsonGet(requestOptions));
    }
    async __getOpenapiSchemaOpenapiJsonGet(requestOptions) {
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "openapi.json"),
            method: "GET",
            headers: _headers,
            queryParameters: requestOptions?.queryParams,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            throw new errors.CambApiError({
                statusCode: _response.error.statusCode,
                body: _response.error.body,
                rawResponse: _response.rawResponse,
            });
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/openapi.json");
    }
}
