// @camb-ai/sdk/dist/api/resources/story/client/Client.js
// This file was auto-generated by Fern from our API Definition.
import { normalizeClientOptionsWithAuth } from "../../../../BaseClient.js";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../../core/headers.js";
import * as core from "../../../../core/index.js";
import * as environments from "../../../../environments.js";
import { handleNonStatusCodeError } from "../../../../errors/handleNonStatusCodeError.js";
import * as errors from "../../../../errors/index.js";
import * as CambApi from "../../../index.js";
export class StoryClient {
    constructor(options) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    /**
     * @param {CambApi.BodyCreateStoryStoryPost} request
     * @param {StoryClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     import { createReadStream } from "fs";
     *     await client.story.createStory({
     *         file: fs.createReadStream("/path/to/your/file"),
     *         source_language: 1
     *     })
     */
    createStory(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__createStory(request, requestOptions));
    }
    async __createStory(request, requestOptions) {
        const _queryParams = {};
        if (request.run_id !== undefined) {
            _queryParams.run_id = request.run_id?.toString() ?? null;
        }
        const _request = await core.newFormData();
        await _request.appendFile("file", request.file);
        _request.append("source_language", request.source_language.toString());
        if (request.title != null) {
            _request.append("title", request.title);
        }
        if (request.description != null) {
            _request.append("description", request.description);
        }
        if (request.narrator_voice_id != null) {
            _request.append("narrator_voice_id", request.narrator_voice_id?.toString() ?? null);
        }
        if (request.folder_id != null) {
            _request.append("folder_id", request.folder_id?.toString() ?? null);
        }
        if (request.chosen_dictionaries != null) {
            for (const _item of request.chosen_dictionaries) {
                _request.append("chosen_dictionaries", _item.toString());
            }
        }
        const _maybeEncodedRequest = await _request.getRequest();
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, mergeOnlyDefinedHeaders({ ..._maybeEncodedRequest.headers }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "story"),
            method: "POST",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            requestType: "file",
            duplex: _maybeEncodedRequest.duplex,
            body: _maybeEncodedRequest.body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/story");
    }
    /**
     * @param {CambApi.BodySetupStoryStorySetupPost} request
     * @param {StoryClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     import { createReadStream } from "fs";
     *     await client.story.setupStory({
     *         file: fs.createReadStream("/path/to/your/file"),
     *         source_language: 1
     *     })
     */
    setupStory(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__setupStory(request, requestOptions));
    }
    async __setupStory(request, requestOptions) {
        const _queryParams = {};
        if (request.run_id !== undefined) {
            _queryParams.run_id = request.run_id?.toString() ?? null;
        }
        const _request = await core.newFormData();
        await _request.appendFile("file", request.file);
        _request.append("source_language", request.source_language.toString());
        if (request.title != null) {
            _request.append("title", request.title);
        }
        if (request.description != null) {
            _request.append("description", request.description);
        }
        if (request.narrator_voice_id != null) {
            _request.append("narrator_voice_id", request.narrator_voice_id?.toString() ?? null);
        }
        if (request.folder_id != null) {
            _request.append("folder_id", request.folder_id?.toString() ?? null);
        }
        if (request.chosen_dictionaries != null) {
            for (const _item of request.chosen_dictionaries) {
                _request.append("chosen_dictionaries", _item.toString());
            }
        }
        const _maybeEncodedRequest = await _request.getRequest();
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, mergeOnlyDefinedHeaders({ ..._maybeEncodedRequest.headers }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "story-setup"),
            method: "POST",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            requestType: "file",
            duplex: _maybeEncodedRequest.duplex,
            body: _maybeEncodedRequest.body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/story-setup");
    }
    /**
     * @param {CambApi.GetStoryStatusStoryTaskIdGetRequest} request
     * @param {StoryClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.story.getStoryStatus({
     *         task_id: "task_id"
     *     })
     */
    getStoryStatus(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getStoryStatus(request, requestOptions));
    }
    async __getStoryStatus(request, requestOptions) {
        const { task_id: taskId, run_id: runId } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `story/${core.url.encodePathParam(taskId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/story/{task_id}");
    }
    /**
     * @param {CambApi.GetStoryRunInfoStoryResultRunIdGetRequest} request
     * @param {StoryClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.story.getStoryRunInfo({
     *         run_id: 1
     *     })
     */
    getStoryRunInfo(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getStoryRunInfo(request, requestOptions));
    }
    async __getStoryRunInfo(request, requestOptions) {
        const { run_id: runId, include_transcript: includeTranscript } = request;
        const _queryParams = {};
        if (includeTranscript !== undefined) {
            _queryParams.include_transcript = includeTranscript?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `story-result/${core.url.encodePathParam(runId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/story-result/{run_id}");
    }
    /**
     * @param {CambApi.GetStoriesRunsResultsStoriesResultsPostRequest} request
     * @param {StoryClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.story.getStoriesRunsResults({
     *         body: {
     *             run_ids: [1]
     *         }
     *     })
     */
    getStoriesRunsResults(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getStoriesRunsResults(request, requestOptions));
    }
    async __getStoriesRunsResults(request, requestOptions) {
        const { run_id: runId, traceparent, body: _body } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, mergeOnlyDefinedHeaders({ traceparent: traceparent !== undefined ? traceparent : undefined }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "stories-results"),
            method: "POST",
            headers: _headers,
            contentType: "application/json",
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            requestType: "json",
            body: _body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/stories-results");
    }
}
