// @camb-ai/sdk/dist/api/resources/textToAudio/client/Client.js
// This file was auto-generated by Fern from our API Definition.
import { normalizeClientOptionsWithAuth } from "../../../../BaseClient.js";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../../core/headers.js";
import * as core from "../../../../core/index.js";
import * as environments from "../../../../environments.js";
import { handleNonStatusCodeError } from "../../../../errors/handleNonStatusCodeError.js";
import * as errors from "../../../../errors/index.js";
import * as CambApi from "../../../index.js";
export class TextToAudioClient {
    constructor(options) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    /**
     * @param {CambApi.CreateTextToAudioRequestPayload} request
     * @param {TextToAudioClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.textToAudio.createTextToAudio({
     *         prompt: "prompt"
     *     })
     */
    createTextToAudio(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__createTextToAudio(request, requestOptions));
    }
    async __createTextToAudio(request, requestOptions) {
        const { run_id: runId, traceparent, ..._body } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, mergeOnlyDefinedHeaders({ traceparent: traceparent !== undefined ? traceparent : undefined }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "text-to-sound"),
            method: "POST",
            headers: _headers,
            contentType: "application/json",
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            requestType: "json",
            body: _body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/text-to-sound");
    }
    /**
     * @param {CambApi.GetTextToAudioStatusTextToSoundTaskIdGetRequest} request
     * @param {TextToAudioClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.textToAudio.getTextToAudioStatus({
     *         task_id: "task_id"
     *     })
     */
    getTextToAudioStatus(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTextToAudioStatus(request, requestOptions));
    }
    async __getTextToAudioStatus(request, requestOptions) {
        const { task_id: taskId, run_id: runId } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `text-to-sound/${core.url.encodePathParam(taskId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/text-to-sound/{task_id}");
    }
    /**
     * Retrieve the Text-to-Audio (TTA) result for a given run.
     *
     * This endpoint validates the provided run ID and retrieves the associated
     * Text-to-Audio output. It supports two output formats:
     * - RAW_BYTES: Streams the audio file in WAV format.
     * - FILE_URL: Returns a pre-signed URL for downloading the audio file.
     *
     * The endpoint ensures that the run type is valid for Text-to-Audio processing
     * and that the storage preferences are applied accordingly.
     *
     * Args:
     *     run_id (int): Unique identifier for the TTA run.
     *     traceparent (Optional[str]): Trace header for distributed tracing.
     *     api_key_obj (dict): API key object containing authentication and storage preferences.
     *     output_type (OutputType): Desired output format. Defaults to RAW_BYTES.
     *
     * Returns:
     *     StreamingResponse | GetTTAOut:
     *         - StreamingResponse: If output_type is RAW_BYTES, a streaming response with the audio in WAV format.
     *         - GetTTAOut: If output_type is FILE_URL, a pre-signed URL to access the audio file.
     *
     * Raises:
     *     HTTPException:
     *         - 400 BAD REQUEST: If the run type is invalid for a TTA run.
     *         - 500 INTERNAL SERVER ERROR: If the audio file cannot be fetched or streamed.
     */
    getTextToAudioResult(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTextToAudioResult(request, requestOptions));
    }
    async __getTextToAudioResult(request, requestOptions) {
        const { run_id: runId, output_type: outputType } = request;
        const _queryParams = {};
        if (outputType != null) {
            _queryParams.output_type = outputType;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `text-to-sound-result/${core.url.encodePathParam(runId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            responseType: "binary-response",
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            throw new errors.CambApiError({
                statusCode: _response.error.statusCode,
                body: _response.error.body,
                rawResponse: _response.rawResponse,
            });
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/text-to-sound-result/{run_id}");
    }
    /**
     * @param {CambApi.GetTextToSoundResultsTextToSoundResultsPostRequest} request
     * @param {TextToAudioClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.textToAudio.getTextToSoundResults({
     *         body: {
     *             run_ids: [1]
     *         }
     *     })
     */
    getTextToSoundResults(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTextToSoundResults(request, requestOptions));
    }
    async __getTextToSoundResults(request, requestOptions) {
        const { run_id: runId, traceparent, body: _body } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, mergeOnlyDefinedHeaders({ traceparent: traceparent !== undefined ? traceparent : undefined }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "text-to-sound-results"),
            method: "POST",
            headers: _headers,
            contentType: "application/json",
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            requestType: "json",
            body: _body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/text-to-sound-results");
    }
}
