// @camb-ai/sdk/dist/api/resources/transcription/client/Client.js
// This file was auto-generated by Fern from our API Definition.
import { normalizeClientOptionsWithAuth } from "../../../../BaseClient.js";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../../core/headers.js";
import * as core from "../../../../core/index.js";
import * as environments from "../../../../environments.js";
import { handleNonStatusCodeError } from "../../../../errors/handleNonStatusCodeError.js";
import * as errors from "../../../../errors/index.js";
import * as CambApi from "../../../index.js";
export class TranscriptionClient {
    constructor(options) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    /**
     * @param {CambApi.BodyCreateTranscriptionTranscribePost} request
     * @param {TranscriptionClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     import { createReadStream } from "fs";
     *     await client.transcription.createTranscription({
     *         language: 1
     *     })
     */
    createTranscription(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__createTranscription(request, requestOptions));
    }
    async __createTranscription(request, requestOptions) {
        const _queryParams = {};
        if (request.run_id !== undefined) {
            _queryParams.run_id = request.run_id?.toString() ?? null;
        }
        const _request = await core.newFormData();
        _request.append("language", request.language.toString());
        if (request.media_file != null) {
            await _request.appendFile("media_file", request.media_file);
        }
        if (request.media_url != null) {
            _request.append("media_url", request.media_url);
        }
        if (request.file != null) {
            await _request.appendFile("file", request.file);
        }
        if (request.audio_url != null) {
            _request.append("audio_url", request.audio_url);
        }
        if (request.project_name != null) {
            _request.append("project_name", request.project_name);
        }
        if (request.project_description != null) {
            _request.append("project_description", request.project_description);
        }
        if (request.folder_id != null) {
            _request.append("folder_id", request.folder_id?.toString() ?? null);
        }
        const _maybeEncodedRequest = await _request.getRequest();
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, mergeOnlyDefinedHeaders({ ..._maybeEncodedRequest.headers }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "transcribe"),
            method: "POST",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            requestType: "file",
            duplex: _maybeEncodedRequest.duplex,
            body: _maybeEncodedRequest.body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/transcribe");
    }
    /**
     * @param {CambApi.GetTranscriptionTaskStatusTranscribeTaskIdGetRequest} request
     * @param {TranscriptionClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.transcription.getTranscriptionTaskStatus({
     *         task_id: "task_id"
     *     })
     */
    getTranscriptionTaskStatus(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTranscriptionTaskStatus(request, requestOptions));
    }
    async __getTranscriptionTaskStatus(request, requestOptions) {
        const { task_id: taskId, run_id: runId } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `transcribe/${core.url.encodePathParam(taskId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/transcribe/{task_id}");
    }
    /**
     * NOTE: This endpoint should be called only by the users to get values for their runs via API.
     * Further we need to validate if the user has access to the run_id, otherwise we should not return the output urls.
     *
     * @param {CambApi.GetTranscriptionResultTranscriptionResultRunIdGetRequest} request
     * @param {TranscriptionClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.transcription.getTranscriptionResult({
     *         run_id: 1
     *     })
     */
    getTranscriptionResult(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTranscriptionResult(request, requestOptions));
    }
    async __getTranscriptionResult(request, requestOptions) {
        const { run_id: runId, word_level_timestamps: wordLevelTimestamps } = request;
        const _queryParams = {};
        if (wordLevelTimestamps !== undefined) {
            _queryParams.word_level_timestamps = wordLevelTimestamps?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `transcription-result/${core.url.encodePathParam(runId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/transcription-result/{run_id}");
    }
    /**
     * @param {CambApi.GetTranscriptionResultsTranscriptionResultsPostRequest} request
     * @param {TranscriptionClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.transcription.getTranscriptionResults({
     *         body: {
     *             run_ids: [1]
     *         }
     *     })
     */
    getTranscriptionResults(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTranscriptionResults(request, requestOptions));
    }
    async __getTranscriptionResults(request, requestOptions) {
        const { run_id: runId, traceparent, body: _body } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, mergeOnlyDefinedHeaders({ traceparent: traceparent !== undefined ? traceparent : undefined }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "transcription-results"),
            method: "POST",
            headers: _headers,
            contentType: "application/json",
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            requestType: "json",
            body: _body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/transcription-results");
    }
}
