// @camb-ai/sdk/dist/api/resources/textToSpeech/client/requests/CreateStreamTtsRequestPayload.js
// This file was auto-generated by Fern from our API Definition.
export var CreateStreamTtsRequestPayload;
(function (CreateStreamTtsRequestPayload) {
    CreateStreamTtsRequestPayload.Language = {
        ArKw: "ar-kw",
        DeCh: "de-ch",
        KoKr: "ko-kr",
        ThTh: "th-th",
        MlIn: "ml-in",
        PtPt: "pt-pt",
        KnIn: "kn-in",
        FiFi: "fi-fi",
        EsMx: "es-mx",
        FrCa: "fr-ca",
        CsCz: "cs-cz",
        PtBr: "pt-br",
        HiIn: "hi-in",
        ArSy: "ar-sy",
        EsUs: "es-us",
        BnBd: "bn-bd",
        JaJp: "ja-jp",
        MrIn: "mr-in",
        ArMa: "ar-ma",
        EsEs: "es-es",
        EnUs: "en-us",
        ZhCn: "zh-cn",
        ElGr: "el-gr",
        PlPl: "pl-pl",
        ArOm: "ar-om",
        FrCh: "fr-ch",
        EnUk: "en-uk",
        EnAu: "en-au",
        ArJo: "ar-jo",
        ArAe: "ar-ae",
        TrTr: "tr-tr",
        ArLy: "ar-ly",
        RuRu: "ru-ru",
        EnIn: "en-in",
        ArYe: "ar-ye",
        ArEg: "ar-eg",
        FrBe: "fr-be",
        TaIn: "ta-in",
        ZhTw: "zh-tw",
        ViVn: "vi-vn",
        BnIn: "bn-in",
        ArSa: "ar-sa",
        DeAt: "de-at",
        PaIn: "pa-in",
        ItIt: "it-it",
        NlNl: "nl-nl",
        ArBh: "ar-bh",
        FrFr: "fr-fr",
        ArQa: "ar-qa",
        UkUa: "uk-ua",
        ArTn: "ar-tn",
        DeDe: "de-de",
        ArXa: "ar-xa",
        ArLb: "ar-lb",
        ZhHk: "zh-hk",
        RoRo: "ro-ro",
        AsIn: "as-in",
        ArIq: "ar-iq",
        NlBe: "nl-be",
        TeIn: "te-in",
        IdId: "id-id",
        ArDz: "ar-dz",
    };
    CreateStreamTtsRequestPayload.SpeechModel = {
        Auto: "auto",
        Mars8: "mars-8",
        Mars8Flash: "mars-8-flash",
        Mars8Instruct: "mars-8-instruct",
        Mars7: "mars-7",
        Mars6: "mars-6",
    };
})(CreateStreamTtsRequestPayload || (CreateStreamTtsRequestPayload = {}));
