// @camb-ai/sdk/dist/api/resources/textToSpeech/client/Client.js
// This file was auto-generated by Fern from our API Definition.
import { normalizeClientOptionsWithAuth } from "../../../../BaseClient.js";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../../core/headers.js";
import * as core from "../../../../core/index.js";
import * as environments from "../../../../environments.js";
import { handleNonStatusCodeError } from "../../../../errors/handleNonStatusCodeError.js";
import * as errors from "../../../../errors/index.js";
import * as CambApi from "../../../index.js";
import { basetenTts } from "../providers/baseten.js";
import { vertexTts } from "../providers/vertex.js";
export class TextToSpeechClient {
    constructor(options) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    /**
     * @throws {@link CambApi.UnprocessableEntityError}
     */
    tts(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__tts(request, requestOptions));
    }
    async __tts(request, requestOptions) {
        // Get provider config from options
        const ttsProvider = this._options.ttsProvider;
        const providerParams = this._options.providerParams;
        // Validate voice_id requirement for default provider
        if (!ttsProvider && !request.voice_id) {
            throw new Error("voice_id is required when using the default Camb.ai provider.");
        }
        // Route to custom provider if configured
        if (ttsProvider === "baseten") {
            return basetenTts(request, providerParams, requestOptions);
        }
        if (ttsProvider === "vertex") {
            return vertexTts(request, providerParams, requestOptions);
        }
        // Default Camb.ai provider
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "tts-stream"),
            method: "POST",
            headers: _headers,
            contentType: "application/json",
            queryParameters: requestOptions?.queryParams,
            requestType: "json",
            body: request,
            responseType: "binary-response",
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/tts-stream");
    }
    /**
     * @param {CambApi.CreateTtsRequestPayload} request
     * @param {TextToSpeechClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.textToSpeech.createTts({
     *         text: "text",
     *         voice_id: 1,
     *         language: 1
     *     })
     */
    createTts(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__createTts(request, requestOptions));
    }
    async __createTts(request, requestOptions) {
        const { run_id: runId, ..._body } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "tts"),
            method: "POST",
            headers: _headers,
            contentType: "application/json",
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            requestType: "json",
            body: _body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/tts");
    }
    /**
     * @param {CambApi.GetTtsResultTtsTaskIdGetRequest} request
     * @param {TextToSpeechClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.textToSpeech.getTtsResult({
     *         task_id: "task_id"
     *     })
     */
    getTtsResult(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTtsResult(request, requestOptions));
    }
    async __getTtsResult(request, requestOptions) {
        const { task_id: taskId, run_id: runId } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `tts/${core.url.encodePathParam(taskId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/tts/{task_id}");
    }
    /**
     * Retrieves the result of a Text To Speech (TTS) run.
     *
     * This endpoint validates the provided `run_id` and fetches the corresponding TTS-generated audio.
     * The user must have valid access to the run. The function supports two output formats:
     * - `RAW_BYTES`: Streams the audio file directly.
     * - `FILE_URL`: Returns a pre-signed URL to download the audio file.
     *
     * Args:
     *     run_id (int): Unique identifier for the TTS run.
     *     api_key_obj (dict): API key object used for authentication and storage preferences.
     *     traceparent (Optional[str]): Traceparent header for distributed tracing.
     *     output_type (OutputType, optional): Determines the output format. Defaults to `RAW_BYTES`.
     *
     * Returns:
     *     StreamingResponse | GetTTSOut:
     *         - If `output_type = RAW_BYTES`: A streaming response containing the TTS-generated audio in FLAC format.
     *         - If `output_type = FILE_URL`: A URL pointing to the stored TTS-generated audio file.
     *
     * Raises:
     *     HTTPException:
     *         - 400 BAD REQUEST if the run ID is invalid or does not belong to a TTS process.
     *         - 500 INTERNAL SERVER ERROR if fetching or streaming the audio fails.
     *
     * Assumptions:
     *     - The user has valid access to the `run_id`.
     *     - The `run_id` corresponds to a valid TTS run.
     *     - There is only **one** dialogue associated with the given `run_id`.
     *
     * @param {CambApi.GetTtsRunInfoTtsResultRunIdGetRequest} request
     * @param {TextToSpeechClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.textToSpeech.getTtsRunInfo({
     *         run_id: 1
     *     })
     */
    getTtsRunInfo(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTtsRunInfo(request, requestOptions));
    }
    async __getTtsRunInfo(request, requestOptions) {
        const { run_id: runId, output_type: outputType } = request;
        const _queryParams = {};
        if (outputType != null) {
            _queryParams.output_type = outputType;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `tts-result/${core.url.encodePathParam(runId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/tts-result/{run_id}");
    }
    /**
     * @param {CambApi.GetTtsResultsTtsResultsPostRequest} request
     * @param {TextToSpeechClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.textToSpeech.getTtsResults({
     *         body: {
     *             run_ids: [1]
     *         }
     *     })
     */
    getTtsResults(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTtsResults(request, requestOptions));
    }
    async __getTtsResults(request, requestOptions) {
        const { run_id: runId, traceparent, body: _body } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, mergeOnlyDefinedHeaders({ traceparent: traceparent !== undefined ? traceparent : undefined }), requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, "tts-results"),
            method: "POST",
            headers: _headers,
            contentType: "application/json",
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            requestType: "json",
            body: _body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return {
                data: _response.body,
                rawResponse: _response.rawResponse,
            };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/tts-results");
    }
    /**
     * @param {CambApi.GetTtsResultDiscordDiscordTtsTaskIdGetRequest} request
     * @param {TextToSpeechClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.textToSpeech.getTtsResultDiscord({
     *         task_id: "task_id"
     *     })
     */
    getTtsResultDiscord(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTtsResultDiscord(request, requestOptions));
    }
    async __getTtsResultDiscord(request, requestOptions) {
        const { task_id: taskId, run_id: runId } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `discord/tts/${core.url.encodePathParam(taskId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/discord/tts/{task_id}");
    }
}
