// @camb-ai/sdk/dist/api/resources/translatedStory/client/Client.js
// This file was auto-generated by Fern from our API Definition.
import { normalizeClientOptionsWithAuth } from "../../../../BaseClient.js";
import { mergeHeaders } from "../../../../core/headers.js";
import * as core from "../../../../core/index.js";
import * as environments from "../../../../environments.js";
import { handleNonStatusCodeError } from "../../../../errors/handleNonStatusCodeError.js";
import * as errors from "../../../../errors/index.js";
import * as CambApi from "../../../index.js";
export class TranslatedStoryClient {
    constructor(options) {
        this._options = normalizeClientOptionsWithAuth(options);
    }
    /**
     * @param {CambApi.CreateTranslationForExistingStoryRequestPayload} request
     * @param {TranslatedStoryClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.translatedStory.createTranslationForExistingStory({
     *         run_id: 1,
     *         target_language: 1
     *     })
     */
    createTranslationForExistingStory(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__createTranslationForExistingStory(request, requestOptions));
    }
    async __createTranslationForExistingStory(request, requestOptions) {
        const { run_id: runId, ..._body } = request;
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `translated-story/${core.url.encodePathParam(runId)}`),
            method: "POST",
            headers: _headers,
            contentType: "application/json",
            queryParameters: requestOptions?.queryParams,
            requestType: "json",
            body: _body,
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "POST", "/translated-story/{run_id}");
    }
    /**
     * @param {CambApi.GetTranslatedStoryStatusTranslatedStoryTaskIdGetRequest} request
     * @param {TranslatedStoryClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.translatedStory.getTranslatedStoryStatus({
     *         task_id: "task_id"
     *     })
     */
    getTranslatedStoryStatus(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTranslatedStoryStatus(request, requestOptions));
    }
    async __getTranslatedStoryStatus(request, requestOptions) {
        const { task_id: taskId, run_id: runId } = request;
        const _queryParams = {};
        if (runId !== undefined) {
            _queryParams.run_id = runId?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `translated-story/${core.url.encodePathParam(taskId)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/translated-story/{task_id}");
    }
    /**
     * @param {CambApi.GetTranslatedStoryRunInfoTranslatedStoryResultRunIdTargetLanguageGetRequest} request
     * @param {TranslatedStoryClient.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link CambApi.UnprocessableEntityError}
     *
     * @example
     *     await client.translatedStory.getTranslatedStoryRunInfo({
     *         run_id: 1,
     *         target_language: 1
     *     })
     */
    getTranslatedStoryRunInfo(request, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTranslatedStoryRunInfo(request, requestOptions));
    }
    async __getTranslatedStoryRunInfo(request, requestOptions) {
        const { run_id: runId, target_language: targetLanguage, include_transcript: includeTranscript } = request;
        const _queryParams = {};
        if (includeTranscript !== undefined) {
            _queryParams.include_transcript = includeTranscript?.toString() ?? null;
        }
        const _authRequest = await this._options.authProvider.getAuthRequest();
        const _headers = mergeHeaders(_authRequest.headers, this._options?.headers, requestOptions?.headers);
        const _response = await core.fetcher({
            url: core.url.join((await core.Supplier.get(this._options.baseUrl)) ??
                (await core.Supplier.get(this._options.environment)) ??
                environments.CambApiEnvironment.Default, `translated-story-result/${core.url.encodePathParam(runId)}/${core.url.encodePathParam(targetLanguage)}`),
            method: "GET",
            headers: _headers,
            queryParameters: { ..._queryParams, ...requestOptions?.queryParams },
            timeoutMs: (requestOptions?.timeoutInSeconds ?? this._options?.timeoutInSeconds ?? 60) * 1000,
            maxRetries: requestOptions?.maxRetries ?? this._options?.maxRetries,
            abortSignal: requestOptions?.abortSignal,
            fetchFn: this._options?.fetch,
            logging: this._options.logging,
        });
        if (_response.ok) {
            return { data: _response.body, rawResponse: _response.rawResponse };
        }
        if (_response.error.reason === "status-code") {
            switch (_response.error.statusCode) {
                case 422:
                    throw new CambApi.UnprocessableEntityError(_response.error.body, _response.rawResponse);
                default:
                    throw new errors.CambApiError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.body,
                        rawResponse: _response.rawResponse,
                    });
            }
        }
        return handleNonStatusCodeError(_response.error, _response.rawResponse, "GET", "/translated-story-result/{run_id}/{target_language}");
    }
}
